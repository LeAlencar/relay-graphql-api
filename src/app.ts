import Koa from "koa";
import cors from "@koa/cors";
import Router from "@koa/router";
import bodyParser from "koa-bodyparser";
import logger from "koa-logger";
import { getGraphQLParameters, processRequest, renderGraphiQL, sendResult, shouldRenderGraphiQL } from 'graphql-helix';
import koaPlayground from 'graphql-playground-middleware-koa';

import { schema } from "./schema/schema";

const app = new Koa();
const router = new Router();

app.use(logger());
app.use(cors());
app.use(bodyParser())

app.on('error', err => {
  // eslint-disable-next-line
  console.log('app error: ', err);
});

router.get('/', async ctx => {
  ctx.body = 'Welcome to your koa server'
})

router.all(
  '/playground',
  koaPlayground({
    endpoint: '/graphql',
  }),
);

router.all('/graphql', async ctx => {
  const request = {
    body: ctx.request.body,
    headers: ctx.req.headers,
    method: ctx.request.method,
    query: ctx.request.query,
  };

  if (shouldRenderGraphiQL(request)) {
    ctx.body = renderGraphiQL({
      subscriptionsEndpoint: '/graphql',
    });
  } else {
    const { operationName, query, variables } = getGraphQLParameters(request);

    const result = await processRequest({
      operationName,
      query,
      variables,
      request,
      schema,
    });

    ctx.respond = false;
    sendResult(result, ctx.res);
  }
});


app.use(router.routes()).use(router.allowedMethods());

export default app;
