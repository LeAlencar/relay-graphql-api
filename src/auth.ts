import jwt from 'jsonwebtoken'
import { jwtSecret } from './config';
import UserModel from './modules/user/UserModel';

export const getUser = async (token: string) => {
  if (!token) return null;

  const decodedToken = jwt.verify(token.substring(4), jwtSecret) as { id: string }

  const user = await UserModel.findOne({_id: decodedToken.id })

  if (!user) return null

  return user

}

export const generateJwtToken = (userId: string) => `JWT ${jwt.sign({ id: userId }, jwtSecret)}`
