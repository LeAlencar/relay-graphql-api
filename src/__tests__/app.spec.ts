import request from 'supertest';

import app from '../app';
import { clearDbAndRestartCounters, connectMongoose, disconnectMongoose } from '../../test';
import { generateJwtToken } from '../auth';
import { createUser } from '../modules/user/fixture/createUser';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

it('should return 200 and return logged user', async () => {
  const user = await createUser();
  const token = generateJwtToken(user);

  const query = `
    query Q {
      user {
        username
        email
      }
    }
  `;

  const variables = {};

  const payload = {
    query,
    variables,
  };

  const response = await request(app.callback())
    .post('/graphql')
    .set({
      Accept: 'application/json',
      'Content-Type': 'application/json',
      Authorization: token,
    })
    .send(JSON.stringify(payload));

  expect(response.body).toMatchSnapshot();
  expect(response.status).toBe(200);
});
