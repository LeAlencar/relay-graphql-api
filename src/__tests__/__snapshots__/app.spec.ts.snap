// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return 200 and return logged user 1`] = `
{
  "data": {
    "user": null,
  },
}
`;
